### PR TITLE
[spirv] Add VK_AMD_shader_early_and_late_fragment_tests support

### PR DESCRIPTION
**Pull request by LLJJDD**

### CODE REVIEW

File: tools/clang/lib/SPIRV/CapabilityVisitor.cpp [hunk 1]
```diff
@@ -608,6 +608,77 @@ bool CapabilityVisitor::visit(SpirvExecutionMode *execMode) {
     addExtension(Extension::KHR_post_depth_coverage,
                  "[[vk::post_depth_coverage]]", execMode->getSourceLocation());
   }
+  if (execMode->getExecutionMode() ==
+      spv::ExecutionMode::EarlyAndLateFragmentTestsAMD) {
+    addExtension(Extension::AMD_shader_early_and_late_fragment_tests,
+                 "[[vk::early_and_late_tests]]", execMode->getSourceLocation());
```
Comment (kuhar) [inline tools/clang/lib/SPIRV/CapabilityVisitor.cpp:614]: same for source location and entry point source location

File: tools/clang/lib/SPIRV/CapabilityVisitor.cpp [hunk 2]
```diff
@@ -608,6 +608,77 @@ bool CapabilityVisitor::visit(SpirvExecutionMode *execMode) {
     addExtension(Extension::KHR_post_depth_coverage,
                  "[[vk::post_depth_coverage]]", execMode->getSourceLocation());
   }
+  if (execMode->getExecutionMode() ==
```
[the remark this answers is not in the document]
Comment (kuhar) [reply tools/clang/lib/SPIRV/CapabilityVisitor.cpp:611]: Or turn the series of `if`s into a `switch`

File: tools/clang/lib/SPIRV/FeatureManager.cpp [hunk 3]
```diff
@@ -203,6 +205,8 @@ const char *FeatureManager::getExtensionName(Extension symbol) {
     return "SPV_AMD_gpu_shader_half_float";
   case Extension::AMD_shader_explicit_vertex_parameter:
     return "SPV_AMD_shader_explicit_vertex_parameter";
+  case Extension::AMD_shader_early_and_late_fragment_tests:
```
Comment (kuhar) [inline tools/clang/lib/SPIRV/FeatureManager.cpp:208]: also here

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 4]
```diff
@@ -11436,6 +11436,70 @@ void SpirvEmitter::processPixelShaderAttributes(const FunctionDecl *decl) {
...
+                                decl->getLocation());
+  }
+
+  // Shaders must not specify more than one of stencil_ref_unchanged_front,
+  // stencil_ref_greater_equal_front, and stencil_ref_less_equal_front.
+  // Shaders must not specify more than one of stencil_ref_unchanged_back,
+  // stencil_ref_greater_equal_back,and stencil_ref_less_equal_back.
+  uint32_t stencilFrontAttrCount = 0, stencilBackAttrCount = 0;
+  if (decl->getAttr<VKStencilRefUnchangedFrontAttr>()) {
+    stencilFrontAttrCount++;
```
Comment (kuhar) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp:11456]: nit: llvm prefers pre-increment: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

Comment (kuhar) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp:11456]: Also below

File: docs/SPIR-V.rst [hunk 5]
```diff
@@ -299,6 +299,7 @@ Supported extensions
 * SPV_GOOGLE_hlsl_functionality1
 * SPV_GOOGLE_user_type
 * SPV_NV_mesh_shader
+* SPV_AMD_shader_early_and_late_fragment_tests
```
Comment (kuhar) [inline docs/SPIR-V.rst:302]: Could you sort this list? This should make it easier to scan for specific extensions

**Review by kuhar — APPROVED** [review]
LGTM

**Comment by LLJJDD** [comment]
Hi @kuhar ,

Is this good to be merged?

**Comment by kuhar** [comment]
@LLJJDD merged

**Comment by LLJJDD** [comment]
> @LLJJDD merged

Thanks!

**Pull request merged**